src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp

```diff
@@ -0,0 +1,233 @@
+  omega_f_of_t_.update(time_of_update, updated_max_deriv, next_expiration_time);
+  update_stored_info();
+}
+
+template <size_t MaxDeriv>
+void QuaternionFunctionOfTime<MaxDeriv>::update_stored_info() const noexcept {
+  const auto& omega_deriv_info = omega_f_of_t_.get_deriv_info();
+
+  ASSERT(omega_deriv_info.size() >= stored_quaternions_and_times_.size(),
+         "There are more stored quaternions than there are stored omegas. "
+         "Currently there are "
+             << omega_deriv_info.size() << " stored omegas and "
+             << stored_quaternions_and_times_.size()
+             << " stored quaternions. There must be at least the same number "
+                "of stored quaternions as omegas otherwise we cannot solve the "
+                "ode and populate the missing quaternions.");
+
+  // If they are already the same size, we don't have to update anything so exit
+  // now
+  if (omega_deriv_info.size() == stored_quaternions_and_times_.size()) {
+    return;
+  }
+
+  // Will store the solution of the ode
+  std::array<DataVector, 1> quaternion_to_integrate;
+  for (size_t i = stored_quaternions_and_times_.size();
+       i < omega_deriv_info.size(); i++) {
+    // Final time, initial time, and quaternion
+    const double t = omega_deriv_info[i].time;
+    const double t0 = omega_deriv_info[i - 1].time;
+    quaternion_to_integrate =
+        stored_quaternions_and_times_[i - 1].stored_quantities;
+
+    solve_quaternion_ode(&quaternion_to_integrate, t, t0);
+
+    // normalize quaternion
+    boost::math::quaternion<double> quat =
+        datavector_to_quaternion(quaternion_to_integrate[0]);
+    normalize_quaternion(make_not_null(&quat));
+
+    stored_quaternions_and_times_.emplace_back(
+        //  FunctionOfTimeHelpers::StoredInfo<1, false>{
+        t, std::array<DataVector, 1>{quaternion_to_datavector(quat)});
+  }
+
+  ASSERT(omega_deriv_info.size() == stored_quaternions_and_times_.size(),
+         "The number of stored omegas must be the same as the number of stored "
+         "quaternions after updating the missing quaternions. Now there are "
+             << omega_deriv_info.size() << " stored omegas and "
+             << stored_quaternions_and_times_.size() << " stored quaternions.");
+}
+
+template <size_t MaxDeriv>
+void QuaternionFunctionOfTime<MaxDeriv>::solve_quaternion_ode(
+    const gsl::not_null<std::array<DataVector, 1>*> quaternion_to_integrate,
+    const double t, const double t0) const noexcept {
+  // lambda that stores the internals of the ode
+  const auto quaternion_ode_system =
+      [this](const std::array<DataVector, 1>& state,
+             std::array<DataVector, 1>& dt_state, const double time) noexcept {
+        const std::array<DataVector, 1> omega = omega_f_of_t_.func(time);
+        dt_state[0] = DataVector{
+            {-0.5 * (state[0][1] * omega[0][0] + state[0][2] * omega[0][1] +
+                     state[0][3] * omega[0][2]),
```

Comment:
Can you integrate the quaternion directly, instead of the DataVector version?  That should be faster and wouldn't require reimplementing quaternion multiplication.

Reply:
No. Unfortunately only fundamental types, spectre vectors, and `std::array`s thereof can be used in odeint. Furthermore, the datatype used in odeint has to be iterable, which `boost::math::quaternion` is specifically not iterable.

Reply:
It works with `boost::numeric::odeint::runge_kutta_dopri5<boost::math::quaternion<double>, double, boost::math::quaternion<double>, double, boost::numeric::odeint::vector_space_algebra>`, although you have to do a slightly sketchy (because it involves no user-defined types) definition of a specialization:
```c++
namespace boost::numeric::odeint {
template <>
struct vector_space_norm_inf<boost::math::quaternion<double>> {
  using result_type = double;
  result_type operator()(const boost::math::quaternion<double>& q) const {
    return sup(q);
  }
};
}  // namespace boost::numeric::odeint                                          
```

Also the errors are a bit different for some reason I don't understand, even without switching to quaternion multiplication.

Reply:
After reading a lot of documentation, yes I now see that works. Added a new commit that implements this.